From 653abf10b5c31cdce1a270b19f5508080766378c Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Marek=20Marczykowski-G=C3=B3recki?=
 <marmarek@invisiblethingslab.com>
Date: Fri, 23 Jun 2023 03:25:28 +0200
Subject: [PATCH] qubes: make 'fwupdmgr get-updates' think it's interactive

When stdout of fwupdmgr is redirected to a pipe, it assumes it's running
non-interactive. This makes it not call
fwupd_client_set_feature_flags(), which results in some (or all) updates not
being listed.

"Fix" this by using new pty for stdout capture, instead of a pipe.
---
 contrib/qubes/src/qubes_fwupdmgr.py | 23 +++++++++++++++++++++--
 1 file changed, 21 insertions(+), 2 deletions(-)

diff --git a/contrib/qubes/src/qubes_fwupdmgr.py b/contrib/qubes/src/qubes_fwupdmgr.py
index a772b940a..887241f22 100755
--- a/contrib/qubes/src/qubes_fwupdmgr.py
+++ b/contrib/qubes/src/qubes_fwupdmgr.py
@@ -7,8 +7,10 @@
 # SPDX-License-Identifier: LGPL-2.1+
 #
 
+import errno
 import json
 import os
+import pty
 import re
 import shutil
 import subprocess
@@ -182,8 +184,25 @@ class QubesFwupdmgr(FwupdHeads, FwupdUpdate, FwupdReceiveUpdates):
     def _get_dom0_updates(self):
         """Gathers information about available updates."""
         cmd_get_dom0_updates = [FWUPDMGR, "--json", "get-updates"]
-        p = subprocess.Popen(cmd_get_dom0_updates, stdout=subprocess.PIPE)
-        self.dom0_updates_info = p.communicate()[0].decode()
+        # connect stdout to a pty, otherwise get-updates works in
+        # non-interactive mode and doesn't list all the updates
+        # (based on FWUUPD_FEATURE_* flags)
+        mstdout, sstdout = pty.openpty()
+        p = subprocess.Popen(cmd_get_dom0_updates, stdout=sstdout)
+        os.close(sstdout)
+        data = b""
+        while True:
+            try:
+                new_data = os.read(mstdout, 4096)
+            except OSError as e:
+                if e.errno == errno.EIO:
+                    # process exited
+                    break
+            if not new_data:
+                break
+            data += new_data
+        self.dom0_updates_info = data.decode()
+        p.wait()
         if p.returncode != 0 and p.returncode != 2:
             raise Exception("fwupd-qubes: Getting available updates failed")
 
-- 
2.39.2

